model_analyzer/config/generate/optuna_run_config_generator.py

```diff
@@ -0,0 +1,302 @@
+        client: TritonClient,
+        model_variant_name_manager: ModelVariantNameManager,
+        search_parameters: SearchParameters,
+        metrics_manager: MetricsManager,
+        seed: Optional[int] = 0,
+    ):
+        """
+        Parameters
+        ----------
+        config: ConfigCommandProfile
+            Profile configuration information
+        gpus: List of GPUDevices
+        models: List of ModelProfileSpec
+            List of models to profile
+        client: TritonClient
+        model_variant_name_manager: ModelVariantNameManager
+        search_parameters: SearchParameters
+            The object that handles the users configuration search parameters
+        """
+        self._config = config
+        self._client = client
+        self._gpus = gpus
+        self._models = models
+        self._search_parameters = search_parameters
+        self._metrics_manager = metrics_manager
+
+        self._model_variant_name_manager = model_variant_name_manager
+
+        self._triton_env = BruteRunConfigGenerator.determine_triton_server_env(models)
+
+        self._num_models = len(models)
+        self._last_measurement: Optional[RunConfigMeasurement] = None
+
+        self._c_api_mode = config.triton_launch_mode == "c_api"
+
+        self._done = False
+
+        if seed is not None:
+            self._sampler = optuna.samplers.TPESampler(seed=seed)
+        else:
+            self._sampler = optuna.samplers.TPESampler()
+
+        self._study = optuna.create_study(
+            study_name=self._models[0].model_name(),
+            direction="maximize",
+            sampler=self._sampler,
+        )
+
+    def _is_done(self) -> bool:
+        return self._done
+
+    def set_last_results(
+        self, measurements: List[Optional[RunConfigMeasurement]]
+    ) -> None:
+        # TODO: TMA-1927: Add support for multi-model
+        if measurements[0] is not None:
+            self._last_measurement = measurements[0]
+        else:
+            self._last_measurement = None
+
+    def get_configs(self) -> Generator[RunConfig, None, None]:
+        """
+        Returns
+        -------
+        RunConfig
+            The next RunConfig generated by this class
+        """
+        default_run_config = self._create_default_run_config()
+        yield default_run_config
+        self._default_measurement = self._last_measurement
+
+        # TODO: TMA-1884: Need a default + config option for trial number
+        n_trials = 20
+        # TODO: TMA-1885: Need an early exit strategy
+        for _ in range(n_trials):
+            trial = self._study.ask()
+            self._create_trial_objectives(trial)
+            run_config = self._create_objective_based_run_config()
+            yield run_config
+            score = self._calculate_score()
+            self._study.tell(trial, score)
+
+    def _create_trial_objectives(self, trial) -> None:
+        # TODO: TMA-1925: Use SearchParameters here
```

Comment:
I see lots of hardcoding limits (for example, instance count 1 to 8). Is SearchParameters going to take care of fixing that?

Reply:
Yes, this is all temporary and SearchParameters will replace this code.